REP-5521 X-Relevant-Roles

dmnjohns opened the pull request:
I'm basing this off of @jorgew's `capture-header-elem` forked branch for the moment (to keep the diff clean), but will move it over to the `api-checker` master branch once that branch is merged.

This PR adds support for relevant roles. Relevant roles are roles which grant a request access to a resource. Technically, these roles are the intersection of roles which belong to the user (i.e., the values of the `X-Roles` header) and roles which are required to access a resource (i.e., the values of `rax:roles`). Relevant roles will be added to the `X-Relevant-Roles` request header.

Commit:
Adding X-RELEVANT-ROLES header

dmnjohns commented on core/src/main/resources/xsl/raxRoles.xsl:
```diff
@@ -113,15 +113,24 @@
+        <xsl:choose>
+            <xsl:when test="not('#all' = $roles)">
+                <rax:captureHeader name="X-RELEVANT-ROLES">
+                    <xsl:attribute name="path" select="string-join(('for $h in req:headers(''X-ROLES'', true()) return if ($h = (''', string-join($roles, ''', '''), ''')) then $h else ()'), '')"/>
```
I'm really not very happy with this. I'm sure there is a better way to express the same thing, but I only know enough XPath to be dangerous.

The thing to point out here is that the `$roles` param is not available in the XPath context, so it can't be inserted directly into the `path` attribute. Instead, `xsl:attribute` is used to populate the `path` attribute. Basically, I'm just using the context that knows about `$roles` to construct a string that will become the value of the `path` attribute. The value of the `path` attribute is an XPath expression which will compare the values from the `X-Roles` header with the values of configured `rax:roles`. Note that the value of the `path` attribute is an XPath expression that will be evaluated in a context which provides the `req:headers` functions.

RackerWilliams replied:
I'd put this into different steps using variables.  For example the first thing you want  to do is put quotes on all the roles right?

````xml
  <xsl:variable name="quoted" as="xsd:string*" select="for $r in $roles return concat(''',$r,''')"/>
````
You actually missed calling ```rax:transformNBSP``` for each role, because you want to list all non-breaking spaces into spaces.  So it may be better to say...

````xml
  <xsl:variable name="quoted" as="xsd:string*" select="for $r in $roles return concat(''',rax:transformNBSP($r),''')"/>
````
 You then want to form the sequence sperated by commas...

````xml
<xsl:variable name="sequence" as="xsd:string"><xsl:value-of select="$quoted" separator=", "/></xsl:variable>
````
This should give you something like 'role1', 'role2', 'role3'.

The other neat trick is that you can avoid using ```xsl:attribute``` by using curly braces in an XML attribute of a template.  This does interpolation of variables or even full expressions. That also avoids things like string-joins and concats..

````xml
    <rax:captureHeader name="X-RELEVANT-ROLES" path="for $h in req:headers('X-ROLES', true) return if ($h = ({$sequence})) then $h else ()" />
````

dmnjohns replied:
Man, I totally remember reading about attribute value templates in the past, but I don't touch XPath very much, and so did not remember them at all.

Anyway, I'll split up the processing a bit and use an attribute value template.

dmnjohns commented on core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRelevantRolesSuite.scala:
```diff
@@ -0,0 +1,116 @@
+import scala.xml.Elem
+
+@RunWith(classOf[JUnitRunner])
+class ValidatorWADLRelevantRolesSuite extends BaseValidatorSuite {
```
This is a pretty light test suite since I only tested the behavior of the few lines of XSL code I added. I don't think it's necessary to stress the state machine since its construction was not touched for this work. Having said that, I know that I'm not familiar enough with `api-checker` to say that these tests cover all of the effects of the code change, so I fully expect someone to call me out and tell me to test certain things that I missed.

RackerWilliams replied:
Make sense, but would like to make sure that this does work will with the mask rax roles feature as this does rewrites on the code you just wrote!

RackerWilliams replied:
Also, you're not really stressing your implementation with this test.  What happens if the role has a single quote in it? Will your code still work?

dmnjohns commented:
I have added some more tests, corrected the bug with handling the `'` character in a role, cleaned up the translation code a bit, and made sure to use the `rax:transformNBSP` function.

dmnjohns committed:
Merge pull request #351 from RackerWilliams/capture-header-elem

Extend Capture Header extension so that it can stand alone
